File: frontend/src/services/userService.ts
import { apiClient } from '@/config/apiClient';
import type { 
  ApiResponse, 
  PageResponse, 
  UserResponse, 
  UserCreateRequest, 
  UserUpdateRequest, 
  ChangePasswordRequest, 
  ResetPasswordRequest,
  TableParams 
} from '@/types';

export const userService = {
  getAll: (params?: TableParams): Promise<ApiResponse<PageResponse<UserResponse>>> => {
    return apiClient.get('/users', { params });
  },

  getMe: (): Promise<ApiResponse<UserResponse>> => {
    return apiClient.get('/users/me');
  },

  getById: (id: number): Promise<ApiResponse<UserResponse>> => {
    return apiClient.get(`/users/${id}`);
  },

  create: (data: UserCreateRequest): Promise<ApiResponse<UserResponse>> => {
    return apiClient.post('/users', data);
  },

  update: (id: number, data: UserUpdateRequest): Promise<ApiResponse<UserResponse>> => {
    return apiClient.put(`/users/${id}`, data);
  },

  updateMe: (data: { fullName: string; phone?: string }): Promise<ApiResponse<UserResponse>> => {
    return apiClient.put('/users/me', data);
  },

  delete: (id: number): Promise<ApiResponse<void>> => {
    return apiClient.delete(`/users/${id}`);
  },

  changePassword: (data: ChangePasswordRequest): Promise<ApiResponse<void>> => {
    return apiClient.post('/users/change-password', data);
  },

  resetPassword: (id: number, data: ResetPasswordRequest): Promise<ApiResponse<void>> => {
    return apiClient.post(`/users/${id}/reset-password`, data);
  },

  uploadAvatar: (file: File): Promise<ApiResponse<UserResponse>> => {
    const formData = new FormData();
    formData.append('file', file);
    return apiClient.post('/users/avatar', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
  },
};

File: backend/src/main/java/com/scim/service/UserService.java
package com.scim.service;

import com.scim.dto.user.request.ChangePasswordRequest;
import com.scim.dto.user.request.ResetPasswordRequest;
import com.scim.dto.user.request.UserCreateRequest;
import com.scim.dto.user.request.UserUpdateRequest;
import com.scim.dto.user.request.UserUpdateMeRequest;
import com.scim.dto.response.PageResponse;
import com.scim.dto.user.response.UserResponse;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

public interface UserService {
    PageResponse<UserResponse> getAll(String search, Pageable pageable);
    UserResponse getById(Long id);
    UserResponse create(UserCreateRequest request);
    UserResponse update(Long id, UserUpdateRequest request);
    UserResponse updateMe(Long userId, UserUpdateMeRequest request);
    void delete(Long id);
    void changePassword(Long userId, ChangePasswordRequest request);
    void resetPassword(Long userId, ResetPasswordRequest request);
    void updateAvatar(Long userId, String avatarUrl);
    UserResponse uploadAvatar(Long userId, MultipartFile file);
}
